// ng-mocks/cjs/lib/common/core.form.js
"use strict";
// tslint:disable variable-name
Object.defineProperty(exports, "__esModule", { value: true });
var AbstractControl;
var DefaultValueAccessor;
var FormControl;
var FormControlDirective;
var NG_ASYNC_VALIDATORS;
var NG_VALIDATORS;
var NG_VALUE_ACCESSOR;
var NgControl;
var NgModel;
try {
    // tslint:disable-next-line no-require-imports no-var-requires
    var module_1 = require('@angular/forms');
    // istanbul ignore else
    if (module_1) {
        AbstractControl = module_1.AbstractControl;
        DefaultValueAccessor = module_1.DefaultValueAccessor;
        FormControl = module_1.FormControl;
        FormControlDirective = module_1.FormControlDirective;
        NG_ASYNC_VALIDATORS = module_1.NG_ASYNC_VALIDATORS;
        NG_VALIDATORS = module_1.NG_VALIDATORS;
        NG_VALUE_ACCESSOR = module_1.NG_VALUE_ACCESSOR;
        NgControl = module_1.NgControl;
        NgModel = module_1.NgModel;
    }
}
catch (e) {
    // nothing to do;
}
exports.default = {
    AbstractControl: AbstractControl,
    DefaultValueAccessor: DefaultValueAccessor,
    FormControl: FormControl,
    FormControlDirective: FormControlDirective,
    NG_ASYNC_VALIDATORS: NG_ASYNC_VALIDATORS,
    NG_VALIDATORS: NG_VALIDATORS,
    NG_VALUE_ACCESSOR: NG_VALUE_ACCESSOR,
    NgControl: NgControl,
    NgModel: NgModel,
};
