// ng-mocks/cjs/lib/common/jit-reflector.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.JitReflector = void 0;
var core_1 = require("@angular/core");
var JitReflector = /** @class */ (function () {
    function JitReflector() {
        this.resolveExternalReference = function (ref) { return ref.runtime; };
        this.reflectionCapabilities = new core_1.ɵReflectionCapabilities();
    }
    JitReflector.prototype.annotations = function (typeOrFunc) {
        return this.reflectionCapabilities.annotations(typeOrFunc);
    };
    JitReflector.prototype.componentModuleUrl = function (type) {
        return "./" + (0, core_1.ɵstringify)(type);
    };
    // This does not exist in Angular 5.1.x but is required to exist in 5.2+
    JitReflector.prototype.guards = function () {
        return {};
    };
    JitReflector.prototype.hasLifecycleHook = function (type, lcProperty) {
        return this.reflectionCapabilities.hasLifecycleHook(type, lcProperty);
    };
    JitReflector.prototype.parameters = function (typeOrFunc) {
        return this.reflectionCapabilities.parameters(typeOrFunc);
    };
    JitReflector.prototype.propMetadata = function (typeOrFunc) {
        return this.reflectionCapabilities.propMetadata(typeOrFunc);
    };
    JitReflector.prototype.shallowAnnotations = function () {
        throw new Error('Not supported in JIT mode');
    };
    JitReflector.prototype.tryAnnotations = function (typeOrFunc) {
        return this.annotations(typeOrFunc);
    };
    return JitReflector;
}());
exports.JitReflector = JitReflector;
