// ng-mocks/cjs/lib/common/mock-control-value-accessor-proxy.js
"use strict";
// tslint:disable variable-name ban-ts-ignore
Object.defineProperty(exports, "__esModule", { value: true });
exports.MockAsyncValidatorProxy = exports.MockValidatorProxy = exports.MockControlValueAccessorProxy = void 0;
var applyProxy = function (proxy, method, value, storage) {
    if (proxy.instance && storage) {
        proxy.instance[storage] = value;
    }
    if (proxy.instance && proxy.instance[method]) {
        return proxy.instance[method](value);
    }
};
var MockControlValueAccessorProxy = /** @class */ (function () {
    function MockControlValueAccessorProxy(target) {
        this.target = target;
    }
    MockControlValueAccessorProxy.prototype.registerOnChange = function (fn) {
        applyProxy(this, 'registerOnChange', fn, '__simulateChange');
    };
    MockControlValueAccessorProxy.prototype.registerOnTouched = function (fn) {
        applyProxy(this, 'registerOnTouched', fn, '__simulateTouch');
    };
    MockControlValueAccessorProxy.prototype.setDisabledState = function (isDisabled) {
        applyProxy(this, 'setDisabledState', isDisabled);
    };
    MockControlValueAccessorProxy.prototype.writeValue = function (value) {
        applyProxy(this, 'writeValue', value);
    };
    return MockControlValueAccessorProxy;
}());
exports.MockControlValueAccessorProxy = MockControlValueAccessorProxy;
var MockValidatorProxy = /** @class */ (function () {
    function MockValidatorProxy(target) {
        this.target = target;
    }
    MockValidatorProxy.prototype.registerOnValidatorChange = function (fn) {
        applyProxy(this, 'registerOnValidatorChange', fn, '__simulateValidatorChange');
    };
    MockValidatorProxy.prototype.validate = function (control) {
        if (this.instance && this.instance.validate) {
            return this.instance.validate(control);
        }
        return null;
    };
    return MockValidatorProxy;
}());
exports.MockValidatorProxy = MockValidatorProxy;
var MockAsyncValidatorProxy = /** @class */ (function () {
    function MockAsyncValidatorProxy(target) {
        this.target = target;
    }
    MockAsyncValidatorProxy.prototype.registerOnValidatorChange = function (fn) {
        applyProxy(this, 'registerOnValidatorChange', fn, '__simulateValidatorChange');
    };
    MockAsyncValidatorProxy.prototype.validate = function (control) {
        if (this.instance && this.instance.validate) {
            var result = this.instance.validate(control);
            return result === undefined ? Promise.resolve(null) : result;
        }
        return Promise.resolve(null);
    };
    return MockAsyncValidatorProxy;
}());
exports.MockAsyncValidatorProxy = MockAsyncValidatorProxy;
