// ng-mocks/cjs/lib/common/ng-mocks-stack.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
var ng_mocks_universe_1 = __importDefault(require("./ng-mocks-universe"));
var stackRoot = { id: {}, level: 'root' };
var stack = ng_mocks_universe_1.default.global.get('reporter-stack') || [__assign({}, stackRoot)];
ng_mocks_universe_1.default.global.set('reporter-stack', stack);
var current = function () { return stack[stack.length - 1]; };
// istanbul ignore next
var listenersPush = (_a = ng_mocks_universe_1.default.global.get('reporter-stack-push')) !== null && _a !== void 0 ? _a : [];
ng_mocks_universe_1.default.global.set('reporter-stack-push', listenersPush);
// istanbul ignore next
var listenersPop = (_b = ng_mocks_universe_1.default.global.get('reporter-stack-pop')) !== null && _b !== void 0 ? _b : [];
ng_mocks_universe_1.default.global.set('reporter-stack-pop', listenersPop);
var stackPush = function () {
    var e_1, _a;
    var id = {};
    ng_mocks_universe_1.default.global.set('reporter-stack-id', id);
    var state = { id: id, level: 'runtime' };
    stack.push(state);
    try {
        for (var listenersPush_1 = __values(listenersPush), listenersPush_1_1 = listenersPush_1.next(); !listenersPush_1_1.done; listenersPush_1_1 = listenersPush_1.next()) {
            var callback = listenersPush_1_1.value;
            callback(state, stack);
        }
    }
    catch (e_1_1) { e_1 = { error: e_1_1 }; }
    finally {
        try {
            if (listenersPush_1_1 && !listenersPush_1_1.done && (_a = listenersPush_1.return)) _a.call(listenersPush_1);
        }
        finally { if (e_1) throw e_1.error; }
    }
};
var stackPop = function () {
    var e_2, _a;
    var state = stack.pop();
    // this code is actually needed for jest tests.
    // istanbul ignore if
    if (stack.length === 0) {
        stack.push((state === null || state === void 0 ? void 0 : state.level) === 'root' ? state : __assign({}, stackRoot));
    }
    // istanbul ignore else
    if (state && state.level !== 'root') {
        try {
            for (var listenersPop_1 = __values(listenersPop), listenersPop_1_1 = listenersPop_1.next(); !listenersPop_1_1.done; listenersPop_1_1 = listenersPop_1.next()) {
                var callback = listenersPop_1_1.value;
                callback(state, stack);
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (listenersPop_1_1 && !listenersPop_1_1.done && (_a = listenersPop_1.return)) _a.call(listenersPop_1);
            }
            finally { if (e_2) throw e_2.error; }
        }
    }
    ng_mocks_universe_1.default.global.set('reporter-stack-id', stack[stack.length - 1].id);
};
// istanbul ignore next
var subscribePush = function (callback) {
    if (listenersPush.indexOf(callback)) {
        listenersPush.push(callback);
    }
    if (stack.length) {
        callback(stack[stack.length - 1], stack);
    }
};
// istanbul ignore next
var subscribePop = function (callback) {
    if (listenersPop.indexOf(callback) === -1) {
        listenersPop.push(callback);
    }
};
// istanbul ignore next
var unsubscribePush = function (callback) {
    var index = listenersPush.indexOf(callback);
    if (index !== -1) {
        listenersPush.splice(index, 1);
    }
};
// istanbul ignore next
var unsubscribePop = function (callback) {
    var index = listenersPop.indexOf(callback);
    if (index !== -1) {
        listenersPop.splice(index, 1);
    }
};
exports.default = {
    current: current,
    stackPop: stackPop,
    stackPush: stackPush,
    subscribePop: subscribePop,
    subscribePush: subscribePush,
    unsubscribePop: unsubscribePop,
    unsubscribePush: unsubscribePush,
};
