// ng-mocks/cjs/lib/mock-service/helper.mock-service.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.registerMockFunction = void 0;
var helper_create_clone_1 = __importDefault(require("./helper.create-clone"));
var helper_create_mock_from_prototype_1 = __importDefault(require("./helper.create-mock-from-prototype"));
var helper_define_property_descriptor_1 = __importDefault(require("./helper.define-property-descriptor"));
var helper_extract_methods_from_prototype_1 = __importDefault(require("./helper.extract-methods-from-prototype"));
var helper_extract_properties_from_prototype_1 = __importDefault(require("./helper.extract-properties-from-prototype"));
var helper_extract_property_descriptor_1 = __importDefault(require("./helper.extract-property-descriptor"));
var helper_mock_1 = __importDefault(require("./helper.mock"));
var helper_mock_function_1 = __importDefault(require("./helper.mock-function"));
var helper_replace_with_mocks_1 = __importDefault(require("./helper.replace-with-mocks"));
var helper_resolve_provider_1 = __importDefault(require("./helper.resolve-provider"));
var helper_use_factory_1 = __importDefault(require("./helper.use-factory"));
// istanbul ignore next
var getGlobal = function () { return window || global; };
// We need a single pointer to the object among all environments.
getGlobal().ngMockshelperMockService = getGlobal().ngMockshelperMockService || {
    mockFunction: helper_mock_function_1.default,
    registerMockFunction: function (func) {
        getGlobal().ngMockshelperMockService.mockFunction.customMockFunction = func;
    },
    createClone: helper_create_clone_1.default,
    createMockFromPrototype: helper_create_mock_from_prototype_1.default,
    definePropertyDescriptor: helper_define_property_descriptor_1.default,
    extractMethodsFromPrototype: helper_extract_methods_from_prototype_1.default,
    extractPropertiesFromPrototype: helper_extract_properties_from_prototype_1.default,
    extractPropertyDescriptor: helper_extract_property_descriptor_1.default,
    mock: helper_mock_1.default,
    replaceWithMocks: helper_replace_with_mocks_1.default,
    resolveProvider: helper_resolve_provider_1.default,
    useFactory: helper_use_factory_1.default,
};
exports.default = (function () { return getGlobal().ngMockshelperMockService; })();
exports.registerMockFunction = getGlobal().ngMockshelperMockService.registerMockFunction;
