// ng-mocks/cjs/lib/mock-builder/mock-builder-stash.js
"use strict";
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MockBuilderStash = void 0;
var core_config_1 = __importDefault(require("../common/core.config"));
var ng_mocks_universe_1 = __importDefault(require("../common/ng-mocks-universe"));
var MockBuilderStash = /** @class */ (function () {
    function MockBuilderStash() {
        this.data = {};
    }
    MockBuilderStash.prototype.backup = function () {
        this.data = {
            builtDeclarations: ng_mocks_universe_1.default.builtDeclarations,
            builtProviders: ng_mocks_universe_1.default.builtProviders,
            cacheDeclarations: ng_mocks_universe_1.default.cacheDeclarations,
            cacheProviders: ng_mocks_universe_1.default.cacheProviders,
            config: ng_mocks_universe_1.default.config,
            configInstance: ng_mocks_universe_1.default.configInstance,
            flags: ng_mocks_universe_1.default.flags,
            touches: ng_mocks_universe_1.default.touches,
        };
        ng_mocks_universe_1.default.builtDeclarations = new Map();
        ng_mocks_universe_1.default.builtProviders = new Map();
        ng_mocks_universe_1.default.cacheDeclarations = new Map();
        ng_mocks_universe_1.default.cacheProviders = new Map();
        ng_mocks_universe_1.default.config = new Map();
        ng_mocks_universe_1.default.configInstance = new Map();
        ng_mocks_universe_1.default.flags = new Set(core_config_1.default.flags);
        ng_mocks_universe_1.default.touches = new Set();
    };
    MockBuilderStash.prototype.restore = function () {
        var e_1, _a;
        try {
            for (var _b = __values(Object.keys(this.data)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var key = _c.value;
                ng_mocks_universe_1.default[key] = this.data[key];
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
    };
    return MockBuilderStash;
}());
exports.MockBuilderStash = MockBuilderStash;
